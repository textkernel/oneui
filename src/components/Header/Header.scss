@import '../../themes/oneui/oneui.scss';

.Header {
    background-color: var(--header-background-color);
    color: var(--header-foreground-color);
    width: 100%;
    height: var(--header-height);
    border-bottom: 1px solid var(--color-light-40);

    &__wrapper {
        box-sizing: border-box;
        padding-top: var(--spacing-normal);
        padding-bottom: var(--spacing-normal);
        height: 100%;
        display: flex;
        justify-content: space-between;
        align-items: center;
        align-content: center;
    }

    &__logo {
        height: var(--header-content-height);
    }
}
