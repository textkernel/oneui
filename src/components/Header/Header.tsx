import * as React from 'react';
import { bem } from '../../utils';
import { BlockWidthRestrictor } from '../WidthRestrictor';
import styles from './Header.scss';

export interface Props extends React.HTMLAttributes<HTMLDivElement> {
    /** supply a logo ready to be rendered or attributes to build a simple one. */
    logo:
        | {
              /** path to the logo source file */
              src: string;
              /** path to where the logo should link to if clicked */
              link?: string;
              /** a name for the logo, it will used as alternative text to the img */
              title?: string;
          }
        | React.ReactElement;
    /** Node(s) to be rendered on the right side of the header */
    children?: ReactNode;
}

const { block, elem } = bem('Header', styles);

export const Header: React.FC<Props> = (props) => {
    const { logo, children, ...rest } = props;

    let renderedLogo: unknown = null;

    if (React.isValidElement(logo)) {
        renderedLogo = React.cloneElement(logo, { ...elem('logo', props) });
    } else if (logo.link) {
        renderedLogo = (
            <a href={logo.link}>
                <img src={logo.src} alt={logo.title} {...elem('logo', props)} />
            </a>
        );
    } else {
        renderedLogo = (
            <img src={logo.src} alt={logo.title || 'our logo'} {...elem('logo', props)} />
        );
    }

    return (
        <header {...rest} {...block(props)}>
            <BlockWidthRestrictor {...elem('wrapper', props)}>
                {renderedLogo}
                <div {...elem('menu', props)}>{children}</div>
            </BlockWidthRestrictor>
        </header>
    );
};

Header.displayName = 'Header';
