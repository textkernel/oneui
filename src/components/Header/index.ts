export { Header, Props as HeaderProps } from './Header';
