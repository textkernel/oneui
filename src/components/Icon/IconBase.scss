@import '../../themes/oneui/oneui';

.IconBase {
    display: inline-flex;
    align-self: center;
    fill: currentColor;

    &--isPrimary {
        fill: var(--color-primary);
    }

    &__svg {
        position: relative;
        top: 0.125em;
        height: 1em;
        transition: fill var(--transition-duration);
    }

    &--margin {
        @each $side in (top, right, bottom, left) {
            &_#{ $side } {
                margin-#{ $side }: 0.25em;
            }
        }
    }
}
