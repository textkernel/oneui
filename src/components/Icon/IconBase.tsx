import React, { forwardRef } from 'react';
import { bem } from '../../utils';
import styles from './IconBase.scss';
import { Context } from '../../constants';

export interface IconProps extends React.HTMLAttributes<HTMLDivElement> {
    /** The icon context (e.g. brand, primary, bad, good etc. - defaults to brand) */
    context?: Context;
    /** Adds margin between a given side of the icon and other content */
    margin?: 'top' | 'right' | 'bottom' | 'left';
    /** Absolute size for this icon (size in pixels, aspect ratio is 1:1).
     If not defined, icon will scale and align itself with text. */
    size?: number;
    /** If true, width will set to be automatic */
    preserveAspectRatio?: boolean;
    /** Optional icon title */
    title?: string;
}

export interface Props extends IconProps {
    /** The SVG content */
    children: NotEmptyReactNode;
    /** The SVG viewbox */
    viewBox: string;
}

const { block, elem } = bem('IconBase', styles);

const adjustSize = (preserveAspectRatio: boolean, size?: number) => {
    if (!size) {
        if (!preserveAspectRatio) {
            return { width: '1em' };
        }
        return { width: 'auto' };
    }

    const adjustedSize = Math.max(0, size);

    return {
        top: 'auto',
        width: preserveAspectRatio ? 'auto' : adjustedSize,
        height: adjustedSize,
    };
};

export const IconBase = forwardRef<HTMLDivElement, Props>(
    (
        { children, context, margin, size, preserveAspectRatio = false, title, viewBox, ...rest },
        ref
    ) => {
        return (
            <div ref={ref} {...rest} {...block({ margin, context, ...rest })}>
                <svg
                    {...elem('svg')}
                    aria-labelledby={title ? 'title' : null}
                    xmlns="http://www.w3.org/2000/svg"
                    viewBox={viewBox}
                    style={adjustSize(!!preserveAspectRatio, size)}
                    role="img"
                >
                    {!!title && <title>{title}</title>}
                    {children}
                </svg>
            </div>
        );
    }
);

IconBase.displayName = 'IconBase';
