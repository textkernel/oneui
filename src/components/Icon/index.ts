export { IconBase } from './IconBase';
export { IconExtract } from './IconExtract';
export { IconHarvester } from './IconHarvester';
export { IconJobfeed } from './IconJobfeed';
export { IconMatch } from './IconMatch';
export { IconSearch } from './IconSearch';
export { IconSourcebox } from './IconSourcebox';
export { IconTextkernel } from './IconTextkernel';
export { LogoTextkernel } from './LogoTextkernel';
