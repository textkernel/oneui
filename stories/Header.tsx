import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { text, withKnobs } from '@storybook/addon-knobs';
import { Header, IconTextkernel } from '@textkernel/oneui';

storiesOf('Molecules/Header', module)
    .addDecorator(withKnobs)
    .add('Header with simple logo', () => (
        <Header
            logo={{
                src: text('Image URL', 'https://www.jobfeed.nl/images/jobfeed-logo.svg'),
                link: text('URL where click on logo will lead', '/'),
                title: text('Alt text for the logo', 'Jobfeed'),
            }}
        >
            {text('Right side', 'This is a placeholder for children')}
        </Header>
    ))
    .add('Header with logo as component', () => (
        <Header logo={<IconTextkernel />}>
            {text('Right side', 'This is a placeholder for children')}
        </Header>
    ));
