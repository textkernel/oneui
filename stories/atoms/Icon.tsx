import * as React from 'react';
import {
    Heading,
    IconExtract,
    IconHarvester,
    IconJobfeed,
    IconMatch,
    IconSearch,
    IconSourcebox,
    IconTextkernel,
    LogoTextkernel,
} from '@textkernel/oneui';
import { HEADING_SIZES } from '../../src/constants';

export default {
    title: 'Atoms/Icons',
    component: IconTextkernel,
};

export const AlignWithText = (args) => (
    <Heading level={HEADING_SIZES[0]}>
        <IconSearch {...args} />
        Some heading
    </Heading>
);
AlignWithText.storyName = 'Align with text';
AlignWithText.args = {
    margin: 'right',
};

export const ExtractIcon = (args) => <IconExtract {...args} />;
ExtractIcon.storyName = 'Extract! icon';
ExtractIcon.args = {
    title: 'Extract!',
};

export const HarvesterIcon = (args) => <IconHarvester {...args} />;
HarvesterIcon.storyName = 'Harvester icon';
HarvesterIcon.args = {
    title: 'Harvester',
};

export const JobfeedIcon = (args) => <IconJobfeed {...args} />;
JobfeedIcon.storyName = 'Jobfeed icon';
JobfeedIcon.args = {
    title: 'Jobfeed',
};

export const MatchIcon = (args) => <IconMatch {...args} />;
MatchIcon.storyName = 'Match! icon';
MatchIcon.args = {
    title: 'Match!',
};

export const SearchIcon = (args) => <IconSearch {...args} />;
SearchIcon.storyName = 'Search! icon';
SearchIcon.args = {
    title: 'Search!',
};

export const SourceboxIcon = (args) => <IconSourcebox {...args} />;
SourceboxIcon.storyName = 'Sourcebox icon';
SourceboxIcon.args = {
    title: 'Sourcebox',
};

export const TextkernelIcon = (args) => <IconTextkernel {...args} />;
TextkernelIcon.storyName = 'Textkernel icon';
TextkernelIcon.args = {
    title: 'Textkernel',
};

export const TextkernelLogo = (args) => <LogoTextkernel {...args} />;
TextkernelLogo.storyName = 'Textkernel logo';
TextkernelLogo.args = {
    title: 'Textkernel logo',
};
