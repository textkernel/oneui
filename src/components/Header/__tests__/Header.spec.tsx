import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Header } from '../Header';
import { IconTextkernel } from '../../Icon';

describe('Header component that renders a website header with a logo on the left and optional children on the right side', () => {
    it('should render correctly with simple logo image', () => {
        const view = render(
            <Header logo={{ src: 'jobfeed-logo.svg' }}>This is a placeholder for children</Header>
        );

        expect(view.container).toMatchSnapshot();
        expect(screen.getByRole('img')).toBeInTheDocument();
        expect(screen.queryByRole('link')).not.toBeInTheDocument();
    });
    it('should render correctly with simple logo image as a link', () => {
        const view = render(
            <Header logo={{ src: 'jobfeed-logo.svg', link: '/' }}>
                This is a placeholder for children
            </Header>
        );

        expect(view.container).toMatchSnapshot();
        expect(screen.getByRole('img')).toBeInTheDocument();
        expect(screen.getByRole('link')).toBeInTheDocument();
    });
    it('should render correctly with logo passed as component', () => {
        const logo = (
            <a href="/test">
                <IconTextkernel aria-label="textkernel icon" />
            </a>
        );
        const view = render(<Header logo={logo}>This is a placeholder for children</Header>);
        const image = screen.getByRole('img');

        expect(view.container).toMatchSnapshot();
        expect(image).toBeInTheDocument();
        expect(image).toHaveAttribute('class', 'IconBase__svg');
        expect(screen.getByRole('link')).toBeInTheDocument();
    });
});
