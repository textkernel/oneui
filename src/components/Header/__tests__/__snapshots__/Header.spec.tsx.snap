// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component that renders a website header with a logo on the left and optional children on the right side should render correctly with logo passed as component 1`] = `
<div>
  <header
    class="Header"
  >
    <div
      class="BlockWidthRestrictor Header__wrapper"
    >
      <a
        class="Header__logo"
        href="/test"
      >
        <div
          class="IconBase"
        >
          <svg
            class="IconBase__svg"
            role="img"
            style="width: 1em;"
            viewBox="0 0 369.7 800"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M334.3,800C203.4,766.9,89.9,681.8,89.9,543.8V248.3H0v-104H89.9V0H223.1V144.2H349.2v104H223.1V554c0,76.5,56.7,109.6,146.6,130.8Z"
            />
          </svg>
        </div>
      </a>
      <div>
        This is a placeholder for children
      </div>
    </div>
  </header>
</div>
`;

exports[`Header component that renders a website header with a logo on the left and optional children on the right side should render correctly with simple logo image 1`] = `
<div>
  <header
    class="Header"
  >
    <div
      class="BlockWidthRestrictor Header__wrapper"
    >
      <img
        alt="our logo"
        class="Header__logo"
        src="jobfeed-logo.svg"
      />
      <div>
        This is a placeholder for children
      </div>
    </div>
  </header>
</div>
`;

exports[`Header component that renders a website header with a logo on the left and optional children on the right side should render correctly with simple logo image as a link 1`] = `
<div>
  <header
    class="Header"
  >
    <div
      class="BlockWidthRestrictor Header__wrapper"
    >
      <a
        href="/"
      >
        <img
          class="Header__logo"
          src="jobfeed-logo.svg"
        />
      </a>
      <div>
        This is a placeholder for children
      </div>
    </div>
  </header>
</div>
`;
